doubly_linked_list/doubly_linked_list.py

```diff
@@ -44,22 +44,55 @@ def __len__(self):
     return self.length
 
   def add_to_head(self, value):
-    pass
+    new_head = ListNode(value)
+    if (self.head == None):
+      self.tail = new_head
+      self.head = new_head
+    else:
+      new_head.next = self.head
+      self.head.prev = new_head
+      self.head = new_head
+    self.length += 1
+      
 
   def remove_from_head(self):
-    pass
-
+      removed = self.head.values
+      if self.head == self.tail:
+        self. head = None
+        self.length -= 1
+      return removed
+        
+      
+    
   def add_to_tail(self, value):
-    pass
+    new_tail = ListNode(value)
+    if (self.tail == None):
+      self.head = new_head
+      self.tail = new_tail #AAAAHHHH
+    else:
+      new_tail.next = self.tail
+      self.tail.next = new_tail
+      self.tail = new_tail
+    self.length += 1
+       
 
   def remove_from_tail(self):
-    pass
+      removed = self.tail.value
+      if self.tail == self.head:
+        self.tail = None
+        self.head = None
+        self.length = 0
+      else:
+        self.tail = self.tail.prev
+        self.tail.next.delete()
+        self.length -= 1
+      return removed
 
```

Comment:
great job, but you might not need to use a delete function. you could just ... let it break off and it will be removed by the garbage collector inside the Python interpreter. Either way, well done!